kroxylicious-filter-test-support/src/main/java/io/kroxylicious/test/record/RecordTestUtils.java

```diff
@@ -295,7 +308,54 @@ public static MemoryRecords memoryRecords(byte magic, long offset, long timestam
     }
 
     private static MemoryRecords memoryRecordsWithoutCopy(byte magic, long offset, long timestamp, byte[] key, byte[] value, Header... headers) {
```

Comment:
I'm not understanding the significance of `WithoutCopy`.  Does appendWithOffset copy the (key, value, header) bytes to the destination buffer before the method returns? Specifically I see the code calling `org.apache.kafka.common.record.DefaultRecord#writeTo` so I think whether it is copied or not is irrelevant.   What am I missing?

Reply:
This was from @tombentley, I think from the comment in one of the usages:
```
// No need to copy the arrays because their contents are written to a ByteBuffer and not retained
```
It's trying to describe something about the way the arrays is used. But as said I don't think it's relevant.

Reply:
(think I'll leave it for a separate round of RecordTestUtils cleanup/improvements)

Reply:
I found some nasty corner cases in testing tests which made me much more defensive about this than I was initially. 

* If the caller passed `byte[]` they could still mutate the contents
* If the caller passed `ByteBuffer`, they could still mutate the contents or the position etc. 

In order to minimise the time wasted investigating tests which are doing weird things I think we want the `Records` returned from this method to be immutable.